### PR TITLE
Refactor motion prediction models

### LINKED ISSUE #194: Refactor prediction/measurement models
- [x] LidarMeasurementModelBase
  - [x] LidarMeasurementModelLikelihood
  - [x] LidarMeasurementModelBeam
- [x] ImuMeasurementModelBase
  - [x] ImuMeasurementModelGravity
- [x] MotionPredictionModelBase
  - [x] MotionPredictionModelDifferentialDrive

### PR DESCRIPTION
**Pull request by nhatao**
Refer #194

### CODE REVIEW

**Comment by at-wat** [comment]
## [#586] PASSED on indigo

<details><summary>All tests passed</summary>


```
build/test_results/mcl_3dl/gtest-test_chunked_kdtree.xml: 1 tests
build/test_results/mcl_3dl/gtest-test_filter.xml: 3 tests
build/test_results/mcl_3dl/gtest-test_imu_measurement_model_gravity.xml: 1 tests
build/test_results/mcl_3dl/gtest-test_motion_prediction_model_differential_drive.xml: 1 tests
build/test_results/mcl_3dl/gtest-test_nd.xml: 2 tests
build/test_results/mcl_3dl/gtest-test_pf.xml: 5 tests
build/test_results/mcl_3dl/gtest-test_point_cloud_random_sampler.xml: 1 tests
build/test_results/mcl_3dl/gtest-test_point_types.xml: 1 tests
build/test_results/mcl_3dl/gtest-test_quat.xml: 4 tests
build/test_results/mcl_3dl/gtest-test_raycast.xml: 3 tests
build/test_results/mcl_3dl/gtest-test_vec3.xml: 5 tests
build/test_results/mcl_3dl/roslint-mcl_3dl.xml: 1 tests
build/test_results/mcl_3dl/rostest-test_tests_expansion_resetting_rostest.xml: 1 tests
build/test_results/mcl_3dl/rostest-test_tests_global_localization_rostest.xml: 1 tests
build/test_results/mcl_3dl/rostest-test_tests_localization_rostest.xml: 1 tests
build/test_results/mcl_3dl/rostest-test_tests_mcl_3dl_compat_rostest.xml: 1 tests
build/test_results/mcl_3dl/rostest-test_tests_tf_rostest.xml: 1 tests
build/test_results/mcl_3dl/rostest-test_tests_transform_rostest.xml: 1 tests
build/test_results/mcl_3dl/rosunit-compare_pose.xml: 1 tests
build/test_results/mcl_3dl/rosunit-compare_tf.xml: 1 tests
build/test_results/mcl_3dl/rosunit-test_expansion_resetting.xml: 1 tests
build/test_results/mcl_3dl/rosunit-test_global_localization.xml: 1 tests
build/test_results/mcl_3dl/rosunit-test_mcl_3dl_compat.xml: 1 tests
build/test_results/mcl_3dl/rosunit-test_transform_failure.xml: 1 tests
Summary: 40 tests, 0 errors, 0 failures, 0 skipped
```
</details>

File: test/src/test_motion_prediction_model_differential_drive.cpp [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+  const mcl_3dl::Vec3 relative_trans_2d(0.5, 0.0, 0.0);
+  const mcl_3dl::Quat relative_quat_2d(mcl_3dl::Vec3(0.0, 0.0, -M_PI / 18.0));
+  const mcl_3dl::State6DOF odom_2d_current(odom_2d_prev.pos_ + odom_2d_prev.rot_ * relative_trans_2d,
+                                           odom_2d_prev.rot_ * relative_quat_2d);
+  predictor.setOdoms(odom_2d_prev, odom_2d_current, 0.1);
+
+  const mcl_3dl::Vec3 relative_trans_s1(0.6, -0.1, 0.0);
+  const mcl_3dl::Quat relative_quat_s1(mcl_3dl::Vec3(0.0, 0.0, -M_PI / 17.0));
+  mcl_3dl::State6DOF s1(odom_2d_prev.pos_ + odom_2d_prev.rot_ * relative_trans_s1,
+                        odom_2d_prev.rot_ * relative_quat_s1);
```
Comment (at-wat) [inline test/src/test_motion_prediction_model_differential_drive.cpp]: Looks better to make the test condition simple to make the test case human validatable.
e.g. split test conditions like:
- odom: only translation, noise: 0
- odom: only rotation, noise: 0
- (optional) odom: geometrically simple combination of rotation and translation, noise: 0
- odom: only translation, with noise value
- ...

Comment (at-wat) [reply test/src/test_motion_prediction_model_differential_drive.cpp]: `s1` is semantically not relative to `odom_2d_prev`.

**Comment by at-wat** [comment]
## [#587] FAILED on indigo

```catkin_make tests``` failed

**Comment by at-wat** [comment]
## [#588] PASSED on indigo

<details><summary>All tests passed</summary>


```
build/test_results/mcl_3dl/gtest-test_chunked_kdtree.xml: 1 tests
build/test_results/mcl_3dl/gtest-test_filter.xml: 3 tests
build/test_results/mcl_3dl/gtest-test_imu_measurement_model_gravity.xml: 1 tests
build/test_results/mcl_3dl/gtest-test_motion_prediction_model_differential_drive.xml: 4 tests
build/test_results/mcl_3dl/gtest-test_nd.xml: 2 tests
build/test_results/mcl_3dl/gtest-test_pf.xml: 5 tests
build/test_results/mcl_3dl/gtest-test_point_cloud_random_sampler.xml: 1 tests
build/test_results/mcl_3dl/gtest-test_point_types.xml: 1 tests
build/test_results/mcl_3dl/gtest-test_quat.xml: 4 tests
build/test_results/mcl_3dl/gtest-test_raycast.xml: 3 tests
build/test_results/mcl_3dl/gtest-test_vec3.xml: 5 tests
build/test_results/mcl_3dl/roslint-mcl_3dl.xml: 1 tests
build/test_results/mcl_3dl/rostest-test_tests_expansion_resetting_rostest.xml: 1 tests
build/test_results/mcl_3dl/rostest-test_tests_global_localization_rostest.xml: 1 tests
build/test_results/mcl_3dl/rostest-test_tests_localization_rostest.xml: 1 tests
build/test_results/mcl_3dl/rostest-test_tests_mcl_3dl_compat_rostest.xml: 1 tests
build/test_results/mcl_3dl/rostest-test_tests_tf_rostest.xml: 1 tests
build/test_results/mcl_3dl/rostest-test_tests_transform_rostest.xml: 1 tests
build/test_results/mcl_3dl/rosunit-compare_pose.xml: 1 tests
build/test_results/mcl_3dl/rosunit-compare_tf.xml: 1 tests
build/test_results/mcl_3dl/rosunit-test_expansion_resetting.xml: 1 tests
build/test_results/mcl_3dl/rosunit-test_global_localization.xml: 1 tests
build/test_results/mcl_3dl/rosunit-test_mcl_3dl_compat.xml: 1 tests
build/test_results/mcl_3dl/rosunit-test_transform_failure.xml: 1 tests
Summary: 43 tests, 0 errors, 0 failures, 0 skipped
```
</details>

File: test/src/test_motion_prediction_model_differential_drive.cpp (hunk 1 above)
Comment (nhatao) [reply test/src/test_motion_prediction_model_differential_drive.cpp]: Tests under various conditions are added. 
As equations become simple when noises are 0, I only added a test in which odometry includes both of rotation and translation when noises are 0. In contrast, when noises are not 0 and odometry has roll and pitch moves, equations become too complicated to validate. So I added tests in which odometry rotates along yaw axis when noises are not 0.

File: test/src/test_motion_prediction_model_differential_drive.cpp [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+  const mcl_3dl::Vec3 state_trans(4.0, 3.0, 2.0);
+  const float state_yaw_angle = 0.3;
+  const mcl_3dl::Quat state_quat(mcl_3dl::Vec3(0.0, 0.0, state_yaw_angle));
+  mcl_3dl::State6DOF state(state_trans, state_quat);
+  state.noise_ll_ = 0.1;
+  state.noise_al_ = 0.2;
+  state.noise_la_ = 0.3;
+  state.noise_aa_ = 0.4;
+
+  const float err_integ_multiply = 0.9;
```
Comment (at-wat) [inline test/src/test_motion_prediction_model_differential_drive.cpp]:
```suggestion
  const float err_integ_multiply = 1.0 - (1.0 / 10.0);  // single step of the exponential smoother
```
would be better for the readability of the test condition

Comment (nhatao) [reply test/src/test_motion_prediction_model_differential_drive.cpp]: Done

**Comment by at-wat** [comment]
## [#589] PASSED on melodic

<details><summary>All tests passed</summary>


```
build/test_results/mcl_3dl/gtest-test_chunked_kdtree.xml: 1 tests
build/test_results/mcl_3dl/gtest-test_filter.xml: 3 tests
build/test_results/mcl_3dl/gtest-test_imu_measurement_model_gravity.xml: 1 tests
build/test_results/mcl_3dl/gtest-test_motion_prediction_model_differential_drive.xml: 4 tests
build/test_results/mcl_3dl/gtest-test_nd.xml: 2 tests
build/test_results/mcl_3dl/gtest-test_pf.xml: 5 tests
build/test_results/mcl_3dl/gtest-test_point_cloud_random_sampler.xml: 1 tests
build/test_results/mcl_3dl/gtest-test_point_types.xml: 1 tests
build/test_results/mcl_3dl/gtest-test_quat.xml: 4 tests
build/test_results/mcl_3dl/gtest-test_raycast.xml: 3 tests
build/test_results/mcl_3dl/gtest-test_vec3.xml: 5 tests
build/test_results/mcl_3dl/roslint-mcl_3dl.xml: 1 tests
build/test_results/mcl_3dl/rostest-test_tests_expansion_resetting_rostest.xml: 1 tests
build/test_results/mcl_3dl/rostest-test_tests_global_localization_rostest.xml: 1 tests
build/test_results/mcl_3dl/rostest-test_tests_localization_rostest.xml: 1 tests
build/test_results/mcl_3dl/rostest-test_tests_mcl_3dl_compat_rostest.xml: 1 tests
build/test_results/mcl_3dl/rostest-test_tests_tf_rostest.xml: 1 tests
build/test_results/mcl_3dl/rostest-test_tests_transform_rostest.xml: 1 tests
build/test_results/mcl_3dl/rosunit-compare_pose.xml: 1 tests
build/test_results/mcl_3dl/rosunit-compare_tf.xml: 1 tests
build/test_results/mcl_3dl/rosunit-test_expansion_resetting.xml: 1 tests
build/test_results/mcl_3dl/rosunit-test_global_localization.xml: 1 tests
build/test_results/mcl_3dl/rosunit-test_mcl_3dl_compat.xml: 1 tests
build/test_results/mcl_3dl/rosunit-test_transform_failure.xml: 1 tests
Summary: 43 tests, 0 errors, 0 failures, 0 skipped
```
</details>

**Pull request merged**